pkg/status/status.go

```diff
@@ -0,0 +1,155 @@
+package status
+
+import (
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	cohelpers "github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/operator-framework/operator-marketplace/version"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const clusterOperatorName = "marketplace-operator"
+
+type status struct {
+	client          *configclient.ConfigV1Client
+	coAPINotPresent bool
+	namespace       string
+	co              *configv1.ClusterOperator
+}
+
+// Status is the interface to report the health of the operator to CVO
+type Status interface {
+	SetFailing(message string)
+	SetAvailable(message string)
+}
+
+// New returns an initialized Status
+func New(client *configclient.ConfigV1Client, namespace string) Status {
+	// Check if the ClusterOperator API is present on the cluster. This is so
+	// that we can continue to work with Kubernetes clusters that don't have
+	// this API.
+	_, err := client.ClusterOperators().Get(clusterOperatorName, v1.GetOptions{})
+	coAPINotPresent := false
+	if meta.IsNoMatchError(err) {
+		log.Warning("ClusterOperator API not present")
+		coAPINotPresent = true
+	}
+
+	return &status{
+		client:          client,
+		coAPINotPresent: coAPINotPresent,
+		namespace:       namespace,
+	}
+}
+
+// SetFailing reports that operator has failed along with the error message
+func (s *status) SetFailing(message string) {
+	s.setStatus(configv1.OperatorFailing, message)
+}
+
+// SetFailing reports that the operator is available to process events
+func (s *status) SetAvailable(message string) {
+	s.setStatus(configv1.OperatorAvailable, message)
+}
+
+// ensureClusterOperator ensures that a ClusterOperator CR is present on the
+// cluster
+func (s *status) ensureClusterOperator() {
+	var err error
+	s.co, err = s.client.ClusterOperators().Get(clusterOperatorName, v1.GetOptions{})
+
+	if err == nil {
+		log.Info("Found existing ClusterOperator")
+		return
+	}
+
+	if err != nil && !errors.IsNotFound(err) {
+		log.Fatalf("Error %v getting ClusterOperator", err)
+	}
+
+	log.Info("Creating ClusterOperator")
+	s.co, err = s.client.ClusterOperators().Create(&configv1.ClusterOperator{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      clusterOperatorName,
+			Namespace: s.namespace,
+		},
+	})
+	if err != nil {
+		log.Fatalf("Error %v creating ClusterOperator", err)
+	}
+}
+
+// setStatus handles setting all the required fields for the given
+// ClusterStatusConditionType
+func (s *status) setStatus(condition configv1.ClusterStatusConditionType, message string) {
+	if s.coAPINotPresent {
+		return
+	}
+	log.Infof("Setting ClusterOperator condition: %s message: %s", condition, message)
+	s.ensureClusterOperator()
```

Comment:
Why run `ensureClusterOperator()` here every time we set the status, rather than once on first startup? It seems a little like unnecessary overhead. I can see the value in some cases, but in ours I would imagine that we just have a single status object that persists and shouldn't lose this reference.

Reply:
I am not sure if CVO modifies the `ClusterOperator` CR. So to be on the safe side I am getting every time we set the status. I see this pattern being followed in the other operators too.